tests/NTest/NTest.md

```diff
@@ -0,0 +1,203 @@
+     -- fails with 'Failed with incorrect error' and 
+     --            'expected errormessage "foo.lua:2: my error" to contain "expected error"'
+``` 
+
+`eq(a, b)` is a helper to deeply compare lua variables. It supports numbers, strings, booleans, nils, functions and tables. It's mostly useful within ok() and nok():
+If the variables are equal it returns `true` else it returns `{msg='<reason>'}` This is recognized by `ok` and `nok` and results in also logging the reason for difference.
+
+``` Lua
+ok(eq(1, 1))
+ok(eq({a='b',c='d'}, {c='d',a='b'})
+ok(eq('foo', 'bar'))     -- will fail
+```
+
+`spy([f])` creates function wrappers that remember each call (arguments, errors) but behaves much like the real function. Real function is optional, in this case spy will return nil, but will still record its calls.
+Spies are most helpful when passing them as callbacks and testing that they were called with correct values.
+
+``` Lua
+local f = spy(function(s) return #s end)
+ok(f('hello') == 5)
+ok(f('foo') == 3)
+ok(#f.called == 2)
+ok(eq(f.called[1], {'hello'})
+ok(eq(f.called[2], {'foo'})
+f(nil)
+ok(f.errors[3] ~= nil)
+```
+
+## Reports
+
+Another useful feature is that you can customize test reports as you need. The default `reports` just more or less prints out a basic report. You can easily override this behavior as well as add any other information you need (number of passed/failed assertions, time the test took etc):
+
+Events are:
+`start`   when testing starts
+`finish`  when all tests have finished
+`begin`   Will be called before each test
+`end`     Will be called after each test
+`pass`    Test has passed
+`fail`    Test has failed with not fulfilled assert (ok, nok, fail)
+`except`  Test has failed with unexpected error
+
+
+``` Lua
+local passed = 0
+local failed = 0
+
+tests.report(function(event, testfunc, msg)
+  if event == 'begin' then
+    print('Started test', testfunc)
+    passed = 0
+    failed = 0
+  elseif event == 'end' then
+    print('Finished test', testfunc, passed, failed)
+  elseif event == 'pass' then
+    passed = passed + 1
+  elseif event == 'fail' then
+    print('FAIL', testfunc, msg)
+    failed = failed + 1
+  elseif event == 'except' then
+    print('ERROR', testfunc, msg)
+  end
+end)
+```
+
+Additionally, you can pass a different environment to keep `_G` unpolluted:
+You need to set it, so the helper functions mentioned above can be added before calling the test function.
+
+``` Lua
+local myenv = {}
+
+tests.report(function() ... end, myenv)
+
+tests.test('Some test', function()
+  myenv.ok(myenv.eq(...))
+  local f = myenv.spy()
+end)
+```
+
+You can set any of the parameters to `nil` to leave the value unchanged. 
+
+
+## Appendix
+
+This Library is for NodeMCU versions Lua 5.1 and Lua 5.3. 
+
+It is based on https://bitbucket.org/zserge/gambiarra and includes bugfixes, substantial extensions of functionality and adaptions to NodeMCU requirements.
```

Comment:
That is a dead end. Maybe link to https://github.com/imolein/gambiarra instead?

Reply:
The source you give is also based on zserges gambiarra. zserge seems to have removed the repo, so I would just link to his account https://bitbucket.org/zserge
But the repo you give has a much better readme, so we might copy some from there.